source/_posts/2018-01-03-using-git-hooks-to-improve-your-day-to-day-workflow.markdown

```diff
@@ -0,0 +1,108 @@
+
+To enable a hook script, put a file in the hooks subdirectory of your `.git` directory that is named appropriately (without any extension) and is executable. From that point forward, it should be called.
+
+In this post we will only be using some *client-side* hooks to execute custom scripts. Next are the hooks and use cases I like the most:
+
+### PRE-COMMIT
+
+The `pre-commit` hook is run first, before you even type in a commit message. It’s used to inspect the snapshot that’s about to be committed, to see if you’ve forgotten something, to make sure tests run, or to examine whatever you need to inspect in the code. Exiting non-zero from this hook aborts the commit, although you can bypass it with `git commit --no-verify`. You can do things like check for code style (run lint or something equivalent), check for trailing whitespace, or check for appropriate documentation on new methods.
+
+{% codeblock lang:bash %}
+#!/usr/bin/env bash
+
+if ! bundle exec rubocop -D -E -S -c .rubocop.yml; then
+  echo "Offences where found. Your commit will not pass the CI."
+  echo "The commit was aborted, but if you really want to commit anyway run: git commit --no-verify -m 'Commit message'"
+  exit 1
+fi
+{% endcodeblock %}
+
+In this example the we use [Rubocop](https://github.com/bbatsov/rubocop) (a linting gem for Ruby) to detect if the code has any issues, if so we let the user know and exit with status code one to prevent the code from being commited.
+
+### COMMIT-MSG
+
+The `commit-msg` hook takes one parameter, which is the path to a temporary file that contains the commit message written by the developer. If this script exits non-zero, Git aborts the commit process, so you can use it to validate your project state or commit message before allowing a commit to go through. You can use it to check that your commit message is conformant to a required pattern or to modify your commit in some useful way.
+
+{% codeblock lang:bash %}
+#!/usr/bin/env bash
+
+if grep -q -i -e "WIP" -e "work in progress" $1; then
+    read -p "You're about to add a WIP commit, do you want to run the CI? [y|n] " -n 1 -r < /dev/tty
+    echo
+    if echo $REPLY | grep -E '^[Nn]$' > /dev/null; then
+        echo "[skip ci]" >> $1
+    fi
+fi
+{% endcodeblock %}
+
+This example searches the commit message entered by the developer for a certain pattern, if the pattern is found it assumes that the commit is a work in progress code and asks the developer to confirm if he wants to run this code in the CI, if the user says *no* then the hook modifies the commit to add a *[skip ci]* label.
```

Comment:
Maybe we should explain what is the meaning of the _[skip ci]_ label, in case reader is not familiar with Travis, CircleCI or anything else supporting this special notation?